external-crates/move/move-vm/profiler/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+        s.split("::").last().unwrap_or(s).to_string()
+    }
+
+    fn is_metered(&self) -> bool {
+        (self.profiles[0].end_value != 0) && (self.start_gas != 0)
+    }
+
+    fn start_gas(&self) -> u64 {
+        self.start_gas
+    }
+
+    fn add_frame(&mut self, frame_name: String) -> u64 {
+        *self
+            .shared
+            .frame_table
+            .entry(frame_name.clone())
+            .or_insert({
+                let val = self.shared.frames.len() as u64;
+                self.shared.frames.push(FrameName { name: frame_name });
+                val as usize
+            }) as u64
+    }
+
+    pub fn open_frame(&mut self, frame_name: String, gas_start: u64) {
+        if !*PROFILER_ENABLED || self.start_gas == 0 {
+            return;
+        }
+
+        let frame_idx = self.add_frame(frame_name.clone());
+        let start = self.start_gas();
+
+        self.profiles[0].events.push(Event {
+            ty: "O".to_string(),
+            frame: frame_idx,
+            at: start - gas_start,
+        });
+    }
+
+    pub fn close_frame(&mut self, frame_name: String, gas_end: u64) {
+        if !*PROFILER_ENABLED || self.start_gas == 0 {
+            return;
+        }
+        let frame_idx = self.add_frame(frame_name);
+        let start = self.start_gas();
+
+        self.profiles[0].events.push(Event {
+            ty: "C".to_string(),
+            frame: frame_idx,
+            at: start - gas_end,
+        });
+        self.profiles[0].end_value = start - gas_end;
+    }
+
+    pub fn to_file(&self) {
+        if !*PROFILER_ENABLED || !self.is_metered() {
+            return;
+        }
+        // Get the unix timestamp
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::UNIX_EPOCH)
+            .expect("Clock may have gone backwards")
+            .as_nanos();
+
+        let mut p = self.config.base_path.clone();
+        p.push(format!("gas_profile_{}_{}.json", self.profile_name(), now));
+        let path_str = p.as_os_str().to_string_lossy().to_string();
+        let mut file = std::fs::File::create(p).expect("Unable to create file");
+
+        let json = serde_json::to_string_pretty(&self).expect("Unable to serialize profile");
+        std::io::Write::write_all(&mut file, json.as_bytes()).expect("Unable to write to file");
+        println!("Gas profile written to file: {}", path_str);
+    }
+}
+
+#[cfg(debug_assertions)]
+impl Drop for GasProfiler {
+    fn drop(&mut self) {
+        let end_gas = self.start_gas() - self.profiles[0].end_value;
+        let mut q = Some(self);
+        profile_close_frame_impl!(&mut q, "root".to_string(), end_gas);
+        profile_dump_file!(q.unwrap());
+        let GasProfiler {
+            exporter: _,
+            name: _,
+            active_profile_index: _,
+            schema: _,
+            shared: _,
+            profiles: _,
+            start_gas: _,
+            config: _,
+        } = self;
```

Comment:
I remember discussing this, but I don't think this will do anything? Since self is `&mut`? When we had discussed this I had thought that `drop` took an owned `self`, but now I am seeing that since it is `&mut self` instead of `self` that is probably how they avoid all sorts of issues like this (since drop is only called on the owned value)

Reply:
Right, thought as much.